working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1595,9 +1634,13 @@ To type check a pattern `p` being matched against a value of type `M`:
 
     [nonnull]: https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md#null-promotion
 
-*   **Constant**: Type check the pattern's value in context type `M`. *The
-    context type comes into play for things like type arguments and
-    int-to-double:*
+*   **Constant**: Nothing to do. When type-checking the constant expression,
+    use no context type.
+
+    *Constant patterns can only appear in refutable contexts, and we don't want
+    the matched value type to cause a coercion that would change the constant
+    value when the intent of the pattern is to see if the matched value _is_
+    that value, not to _make_ it so. In other words, this does not match:*
```

Comment:
FWIW, we currently do do this conversion (it immediately results in an error, but we do it).

Reply:
I'll follow up on [the bug](https://github.com/dart-lang/language/issues/2587), but I'm generally OK with relying on the existing equality behavior around doubles for matching as well. The set of users who actually care about NaN and distinguishing negative and positive zero is *very small* and they generally know that they have to be careful.

So I don't think numbers are a strong signal for deciding how we want to handle context types for constant patterns.

I know we've gone back and forth on this a number of times. My recollection is that after talking to @leafpetersen a bunch, I was pretty well convinced that *not* using the context type was the right choice, though I can't remember the arguments right now. Unless there's a compelling reason outside of numbers, I think it's probably best to leave the current behavior as specified.

Reply:
I'm not sure whether it's a good example, but consider:
```dart
switch (someConstantSet) {
  case const {1, 2}: ...
  case const {1}: ...
  case const {}: ...
}
```
Here the final `const{}` is actually a map. If we had used the matchee type as the context type of the constant, then it would be a set. (In practice, if you have a bunch of constant sets that you are looking for, they'll likely be named constants anyway.)

So, probably not going to happen in real code, but I see no *harm* in providing a context type. If we introduce more context-dependent literals in the future, we might get more examples.

It just feels *right* to provide the context type. Throwing away available information unnecessarily feels like something that will come back and haunt us later. Very much a gut feeling, nothing practical and concrete I can point to in the current language.

Reply:
OK, using the context type SGTM.